lib/AST/ASTContext.cpp

```diff
@@ -2539,91 +2540,126 @@ get(GenericTypeDecl *TheDecl, Type Parent, const ASTContext &C) {
+
+  // Now, map the declaration's own generic parameters to
+  // the provided arguments.
+  {
+    const auto GenParams = Sig->getInnermostGenericParams();
+    for (const auto I: indices(GenericArgs)) {
+      Subs[GenParams[I]->getCanonicalType()
+               ->castTo<GenericTypeParamType>()] = GenericArgs[I];
+    }
+  }
+
+  // If we haven't been given a module for conformance lookup,
+  // use the parent module of the referenced decl.
+  if (M == nullptr)
+    M = TheDecl->getParentModule();
+
+  // Form the substitution map.
+  const auto SubMap = SubstitutionMap::get(
+      Sig, QueryTypeSubstitutionMap{Subs}, LookUpConformanceInModule(M));
+
+  return get(TheDecl, Parent, SubMap);
+}
+
+BoundGenericType *BoundGenericType::get(NominalTypeDecl *TheDecl,
+                                        Type Parent,
+                                        SubstitutionMap Substitutions) {
+  assert(!Substitutions.empty() && "must have non-empty substitutions");
+  assert(TheDecl->isGeneric() && "must be a generic type decl");
   assert((!Parent || Parent->is<NominalType>() ||
           Parent->is<BoundGenericType>() ||
           Parent->is<UnboundGenericType>()) &&
          "parent must be a nominal type");
 
   ASTContext &C = TheDecl->getDeclContext()->getASTContext();
   llvm::FoldingSetNodeID ID;
-  BoundGenericType::Profile(ID, TheDecl, Parent, GenericArgs);
+  BoundGenericType::Profile(ID, TheDecl, Parent, Substitutions);
   RecursiveTypeProperties properties;
   if (Parent) properties |= Parent->getRecursiveProperties();
-  for (Type Arg : GenericArgs) {
-    properties |= Arg->getRecursiveProperties();
-  }
+  for (const auto substGP : Substitutions.getReplacementTypes())
+    properties |= substGP->getRecursiveProperties();
 
   auto arena = getArena(properties);
 
   void *InsertPos = nullptr;
   if (BoundGenericType *BGT =
-        C.getImpl().getArena(arena).BoundGenericTypes.FindNodeOrInsertPos(ID,
+        C.getImpl().getArena(arena).BoundGenericTypes1.FindNodeOrInsertPos(ID,
                                                                      InsertPos))
     return BGT;
 
-  bool IsCanonical = !Parent || Parent->isCanonical();
-  if (IsCanonical) {
-    for (Type Arg : GenericArgs) {
-      if (!Arg->isCanonical()) {
-        IsCanonical = false;
-        break;
-      }
-    }
-  }
+  const bool IsCanonical = [&] {
+    bool IsCanonical = !Parent || Parent->isCanonical();
+    for (const auto Arg : Substitutions.getReplacementTypes())
+      if (!Arg->isCanonical())
+        return false;
+
+    return IsCanonical;
+  }();
 
   BoundGenericType *newType;
   if (auto theClass = dyn_cast<ClassDecl>(TheDecl)) {
-    auto sz = BoundGenericClassType::totalSizeToAlloc<Type>(GenericArgs.size());
+    auto sz = BoundGenericClassType::totalSizeToAlloc<SubstitutionMap>(1);
     auto mem = C.Allocate(sz, alignof(BoundGenericClassType), arena);
     newType = new (mem) BoundGenericClassType(
```

Comment:
`BoundGeneric*Type` no longer have any storage of their own. Is it okay of I lift `mem` out of the blocks here?